### PR TITLE
Automotive TrajectoryAgents

### PR DESCRIPTION
**Pull request by jadecastro**
Toward #8529.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8725)

### CODE REVIEW

**Comment by jadecastro** [comment]
+@stonier feature review.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8725#-:-LBOTf1d7sTGCv4YXTkb:b-kcy23c)*

File: automotive/trajectory_agent.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+/// taking as input an AgentData structure and an AgentTrajectory.
+///
+/// output port 0:
+/// * position: x, y, heading;
+///   heading is 0 rad when pointed +x, pi/2 rad when pointed +y;
+///   heading is defined around the +z axis, so positive-turn-left
+/// * velocity
+///   (OutputPort getter: raw_pose_output())
+///
+/// output port 1: A PoseVector containing X_WC, where C is the car frame.
```
Comment (stonier) [inline automotive/trajectory_agent.h]: Agent, !Car -> X_WA?

**Comment by stonier** [comment]
:LGTM: pending what looks like a flaky CI

---







*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8725#-:-LBdqQLwEW6KRYCybR4x:bxmjoxh)*

**Comment by jadecastro** [comment]
+@sherm1 for platform review please.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8725#-:-LBfFcJa0rW1DqR7i2TT:b6vi582)*

**Comment by jadecastro** [comment]
BTW, I’m not entirely sure what to make of the CI error. It looks like a pybind thing but not sure. 
@ericcousineau-tri do you have any thoughts on how to fix this?

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8725#-:-LBfG0KX81eLs6gFAhyw:bhdld1f)*

**Comment by EricCousineau-TRI** [comment]
I don't think it's `pybind`-related (whew!), but rather related to this PR: https://github.com/RobotLocomotion/drake/pull/8664

Per the fixes Jeremy submitted in https://github.com/RobotLocomotion/drake/pull/8726, I think you'd add the target to the dependency list:
```
error: Missing deps in //automotive's drake_cc_package_library.
note: In the automotive/BUILD.bazel rule for drake_cc_package_library, add the following lines to the deps:
        ":trajectory_agent",
```

**Comment by jadecastro** [comment]
Ah, :facepalm:, thanks. I should have looked at the updated BUILD file.  It's clear to me now.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8725#-:-LBfcWH_6JtPCobJ4-Vb:b-lco488)*

**Comment by jadecastro** [comment]
Comments addressed.

---

Review status: 3 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[automotive/trajectory_agent.h, line 34 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBhkU0568KL36FiMJ3d:-LBsJ_hq0w3mGNulzggh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/eac0f7270e55829edf2654161135b64b09b63694/automotive/trajectory_agent.h#L34)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider saying this is a SimpleCarState.
</blockquote></details>

Done.

---

*[automotive/trajectory_agent.h, line 38 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBhjyct63h092Wo5Neb:-LBsJeNh7bsWmx1b9-cb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/eac0f7270e55829edf2654161135b64b09b63694/automotive/trajectory_agent.h#L38)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW is this ẋ ẏ heading_dot? (consider clarifying)
</blockquote></details>

Done.

---

*[automotive/trajectory_agent.h, line 45 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBhmNdt67B8r5h9dHUP:-LBsJgMo-S5ELG0LrXTZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/eac0f7270e55829edf2654161135b64b09b63694/automotive/trajectory_agent.h#L45)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW this is not correct (and I see FrameVelocity's documentation is wrong too :( ). What you are outputting here is the spatial velocity of A in W, which we write V_WA. That is not the same thing as Xdot_WA, which literally means d/dt X_WA. The derivative of a vector or matrix quantity is an identically-shaped quantity whose elements are the time derivatives of the original elements. That is definitely not true here since (1) a PoseVector has 7 elements while FrameVelocity has 6, and (I hate to say this) (2) the rotational and translational elements are flipped around (!!).
</blockquote></details>

Done.

---

*[automotive/trajectory_agent.h, line 47 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBhoiQx9BGV148MzeDF:-LBsJhWF28C5HNHeW43P:b-j737vf) ([raw file](https://github.com/robotlocomotion/drake/blob/eac0f7270e55829edf2654161135b64b09b63694/automotive/trajectory_agent.h#L47)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW why do you use FrameVelocity here rather than SpatialVelocity? I notice you use PoseVelocity elsewhere and that uses SpatialVelocity.
</blockquote></details>

I'm using FrameVelocity simply because I want all outputs to be Vector-based, as opposed to Abstract outputs. SpatialVelocity does not specialize BasicVector, so instead I'm using its BasicVector container (FrameVelocity).

---

*[automotive/trajectory_agent.h, line 66 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBhnj9F4L9PMypS39_0:-LBsKv1T8PWkZKT47nit:b-lml4ol) ([raw file](https://github.com/robotlocomotion/drake/blob/eac0f7270e55829edf2654161135b64b09b63694/automotive/trajectory_agent.h#L66)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW does this format correctly in doxygen?
</blockquote></details>

Here's the result:


---

*[automotive/trajectory_agent.h, line 76 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBhnlfNDzCnJ_sArMwv:-LBsLbDh1kUEG-1i-JD-:b1hslh8) ([raw file](https://github.com/robotlocomotion/drake/blob/eac0f7270e55829edf2654161135b64b09b63694/automotive/trajectory_agent.h#L76)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW you need an `@name` here for grouping to work in doxygen.

Please generate the doxygen for this PR and inspect it to make sure it is formatted as you expected (and confirm here that it is).
</blockquote></details>

Done. The `@name` keyword fixed the formatting. Thanks!

---

*[automotive/trajectory_agent.h, line 103 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBhp3ddBGMqx1N-Efvw:-LBsLqUQ8z9qHBQmA6VF:b-6ctxyj) ([raw file](https://github.com/robotlocomotion/drake/blob/eac0f7270e55829edf2654161135b64b09b63694/automotive/trajectory_agent.h#L103)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW could this be SpatialVelocity? Then you can just copy it directly out of PoseVelocity.
</blockquote></details>

(see above for the `FrameVelocity` rationale)

---

*[automotive/trajectory_agent.cc, line 40 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBhq2YU7C-4n5Ql52ut:-LBsM6L29qGLaMOUcksg:belnatb) ([raw file](https://github.com/robotlocomotion/drake/blob/eac0f7270e55829edf2654161135b64b09b63694/automotive/trajectory_agent.cc#L40)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I see that PoseVelocity isn't templatized. Does that mean the derivatives produced when T=AutoDiff are wrong? If so, please document. Otherwise please add a comment (not sure if this is the right place) explaining why everything is still good.
</blockquote></details>

Done. (I've added a note within the class documentation to clarify.)

---

*[automotive/test/trajectory_agent_test.cc, line 57 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBhr6_C4VAyLM9xaNVv:-LBsMHNT85PcUTWdaOD1:bio5u6b) ([raw file](https://github.com/robotlocomotion/drake/blob/eac0f7270e55829edf2654161135b64b09b63694/automotive/test/trajectory_agent_test.cc#L57)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI spacing seems odd.
</blockquote></details>

Done. (I think this was clang-format's fault)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8725#-:-LBsM_jj1fnWM6z4RQ8C:bel95ya)*

**Comment by sherm1** [comment]
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[automotive/trajectory_agent.h, line 41 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBsVRtV9T-owVhL_NbH:-LBsVRtV9T-owVhL_NbI:beob9au) ([raw file](https://github.com/robotlocomotion/drake/blob/ffb73a7c2ae663eb1f093cc9f6408b12cce5274f/automotive/trajectory_agent.h#L41)):*
> ```objc
> ///   heading is 0 rad when pointed +x, pi/2 rad when pointed +y;
> ///   heading is defined around the +z axis, positive-left-turn.
> /// * speed: ṡ = √{ẋ² + ẏ²}
> ```

BTW ṡ -> s ? (Looks more like an acceleration if it is speed_dot.)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8725)*

**Comment by jadecastro** [comment]
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[automotive/trajectory_agent.h, line 41 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8725#-LBsVRtV9T-owVhL_NbH:-LBtzboC6ZriokoWqvTn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ffb73a7c2ae663eb1f093cc9f6408b12cce5274f/automotive/trajectory_agent.h#L41)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW ṡ -> s ? (Looks more like an acceleration if it is speed_dot.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8725)*

**Comment by jadecastro** [comment]
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-cache please

**Pull request merged**